### PR TITLE
Implemented SAMPLED_ADDMM

### PR DESCRIPTION
**Pull request by mattbahr**
Relates to #56311 

I have the operation implemented for cpu/gpu, but not for tpu/xla. If someone has time, I'd appreciate a quick review to make sure I have the right idea as well as any advice for where to go from here. Currently I only have the op registered for the float type.

### CODE REVIEW

File: tensorflow/python/ops/math_ops.py [hunk 1]
```diff
@@ -4857,6 +4857,68 @@ def sparse_segment_sum(
...
+@tf_export("sparse.sampled_addmm")
+def sampled_addmm(indices,
+    values,
+    dense_shape,
+    mat1,
+    mat2,
+    beta=1.0,
+    alpha=1.0
+):
+  sparse_rank = len(dense_shape)
```
Comment (cantonios) [inline tensorflow/python/ops/math_ops.py:4869]: This won't always work if `dense_shape` is a dynamic shape obtained from a tensor, e.g. via `tf.shape`.

File: tensorflow/python/ops/math_ops.py [hunk 2]
```diff
@@ -4857,6 +4857,68 @@ def sparse_segment_sum(
...
+  sparse_num_batches = dense_shape[0] if sparse_rank == 3 else 1
```
Comment (cantonios) [inline tensorflow/python/ops/math_ops.py:4870]: You _could_ have this work for multiple batch dimensions, using the last two as the matrix dimensions.

File: tensorflow/python/ops/math_ops.py [hunk 3]
```diff
@@ -4857,6 +4857,68 @@ def sparse_segment_sum(
...
+  sparse_num_rows = dense_shape[0 if sparse_rank == 2 else 1]
+  sparse_num_cols = dense_shape[1 if sparse_rank == 2 else 2]
+
+  values_shape = values.get_shape()
+  values_shape_list = values_shape.as_list()
```
Comment (cantonios) [inline tensorflow/python/ops/math_ops.py:4875]: I think this will only work in Eager mode.

File: tensorflow/python/ops/math_ops.py [hunk 4]
```diff
@@ -4857,6 +4857,68 @@ def sparse_segment_sum(
...
+  mat1_shape = mat1.get_shape()
+  mat1_shape_list = mat1_shape.as_list()
+  mat2_shape = mat2.get_shape()
+  mat_rank = mat1_shape.ndims
+
+  mat_num_batches = mat1_shape_list[0] if mat_rank == 3 else 1
+  mat1_num_rows = mat1_shape_list[0 if mat_rank == 2 else 1]
+  mat2_num_cols = mat2_shape.as_list()[1 if mat_rank == 2 else 2]
+
+  if values_shape.ndims != sparse_rank - 1:
```
Comment (cantonios) [inline tensorflow/python/ops/math_ops.py:4888]: These shape validation checks are probably best done by the kernel itself, or its shape function when registering the op (if ranks are known).  You could also let the kernel itself determine the number of batch dimensions, etc.

The python code in TF is mainly used to set up a graph of execution, and may not have full tensor size information at the time (e.g. if the compiled graph is meant to work for tensors of varying sizes).  But by the time the graph is executed and we enter the kernel, the full size of the inputs will be known.

File: tensorflow/core/ops/math_ops.cc [hunk 5]
```diff
@@ -978,6 +978,17 @@ REGISTER_OP("SparseMatMul")
...
+REGISTER_OP("SampledADDMM")
+    .Input("indices: int32")
+    .Input("values: T")
+    .Input("dense_shape: int32")
+    .Input("mat1: T")
+    .Input("mat2: T")
+    .Attr("T: type")
+    .Attr("beta: float = 1.0")
+    .Attr("alpha: float = 1.0")
+    .Output("output: T");
```
Comment (cantonios) [inline tensorflow/core/ops/math_ops.cc:990]: Here you may want to set a shape function that tries to validate the input shapes (if known).

File: tensorflow/core/kernels/sampled_addmm_op.cc [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+template <typename T>
+struct SampledADDMMFunctor<CPUDevice, T> {
+  static Status Compute(OpKernelContext* ctx,
+                     const Tensor& indices_t, const Tensor& values_t, 
+                     const Tensor& mat1, const Tensor& mat2, 
+                     const int32_t batch_size, const T beta_, const T alpha_,
+                     const int32_t mat1_num_rows, const int32_t mat1_num_cols,
+                     const int32_t mat2_num_rows, const int32_t mat2_num_cols,
+                     const int32_t mat_num_batches, Tensor* out) {
+    auto mat1_ptr = mat1.flat<T>().data();
```
Comment (cantonios) [inline tensorflow/core/kernels/sampled_addmm_op.cc:23]: Please avoid using `auto` when the type is known and simple (e.g. `T*`).

**Review by cantonios — CHANGES_REQUESTED** [review]

**Comment by cantonios** [comment]
You might be able to do this with existing ops:
- `tf.gather_nd` to extract the appropriate rows/columns of `mat_a` and `mat_b`.
- `tf.matmul` to do the dot-product
- scale and sum to the sparse input values

The benefit of doing it this way is XLA should just work out of the box, and you wouldn't need to support a new op.

**Comment by mattbahr** [comment]
@cantonios Thanks for the feedback! Will look more into it over the next couple days. Working on a shape function now.

File: tensorflow/python/ops/math_ops.py (hunk 1 above)
Comment (mattbahr) [reply tensorflow/python/ops/math_ops.py:4869]: I guess we could probably assume the dense shape based on the shape of the product of the matrix multiplication. So we could probably get rid of this `dense_shape` input altogether

Comment (cantonios) [reply tensorflow/python/ops/math_ops.py:4869]: You may want to verify they are the same.  The input here is essentially a `SparseTensor` broken down into components.  I would keep it as is.

File: tensorflow/core/kernels/sampled_addmm_op.cc [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                     const int32_t mat2_num_rows, const int32_t mat2_num_cols,
+                     const int32_t mat_num_batches, Tensor* out) {
+    const T* mat1_ptr = mat1.flat<T>().data();
+    const T* mat2_ptr = mat2.flat<T>().data();
+    const T* values_ptr = values_t.flat<T>().data();
+    const int32_t* indices_ptr = indices_t.flat<int32_t>().data();
+    const int32_t* ds_ptr = dense_shape_t.flat<int32_t>().data();
+
+    int32_t ds_num_rows = ds_ptr[0];
+    int32_t ds_num_cols = ds_ptr[1];
```
Comment (mattbahr) [inline tensorflow/core/kernels/sampled_addmm_op.cc:30]: Getting a segmentation fault when I perform this same operation from the gpu implementation of this function

**Comment by mattbahr** [comment]
> You might be able to do this with existing ops:
> 
> * `tf.gather_nd` to extract the appropriate rows/columns of `mat_a` and `mat_b`.
> * `tf.matmul` to do the dot-product
> * scale and sum to the sparse input values
> 
> The benefit of doing it this way is XLA should just work out of the box, and you wouldn't need to support a new op.

It seems like `tf.gather_nd` will only gather slices horizontally, so I had to transpose `mat_b` to extract the columns. From there instead of `tf.matmul` I just did an element-wise multiplication and then used `tf.math.reduce_sum` to get the dot product. But I think you're right in that we can do this without a new op. I'll see about just implementing it in `math_ops.py`.

```>>> import tensorflow as tf
2024-01-05 13:56:56.117075: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
>>> indices = tf.constant([[[0,1],[1,0]],[[0,0],[1,1]]])
2024-01-05 13:57:04.040611: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:962] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-05 13:57:04.096683: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:962] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-05 13:57:04.096895: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:962] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-05 13:57:04.098041: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:962] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-05 13:57:04.098207: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:962] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-05 13:57:04.098354: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:962] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-05 13:57:04.170050: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:962] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-05 13:57:04.170255: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:962] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-05 13:57:04.170419: I external/local_xla/xla/stream_executor/cuda/cuda_executor.cc:962] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355
2024-01-05 13:57:04.170549: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1929] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 5734 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 3070, pci bus id: 0000:2d:00.0, compute capability: 8.6
>>> values = tf.constant([[0.12,0.0256],[0.54,0.77]])
>>> mat1 = tf.constant([[[0.7,0.094,1.043],[1.4,0.26,6.03]],[[0.842,0.439,0.218],[0.666,0.988,0.407]]])
>>> mat2 = tf.constant([[[0.076,0.99],[1.23,1.04],[0.98,0.326]],[[0.309,0.208],[0.557,0.4681],[0.6,0.742]]])
>>> alpha = 0.43
>>> beta = 0.77
>>> indices_dims = indices.get_shape().ndims
>>> row_indices = tf.slice(indices, [0] * indices_dims, [-1] * (indices_dims - 1) + [1])
>>> col_indices = tf.slice(indices, [0] * (indices_dims - 1) + [1], [-1] * (indices_dims - 1) + [1])
>>> rows = tf.gather_nd(mat1, row_indices, batch_dims=indices_dims-2)
>>> mat2_transpose = tf.transpose(mat2, perm=[0] * (indices_dims - 2) + [indices_dims - 1] + [indices_dims - 2])
>>> cols = tf.gather_nd(mat2_transpose, col_indices, batch_dims=indices_dims - 2)
>>> product = rows * cols
>>> dot = tf.reduce_sum(product, axis=-1)
>>> res = alpha * dot + beta * values
>>> print(res)
tf.Tensor(
[[0.57863456 2.7440202 ]
 [0.68906546 0.981192  ]], shape=(2, 2), dtype=float32)```

File: tensorflow/python/ops/math_ops.py [hunk 8]
```diff
@@ -4857,6 +4857,42 @@ def sparse_segment_sum(
         name=name,
     )
 
+@tf_export("sparse.sampled_addmm")
+def sampled_addmm(indices,
+    values,
+    dense_shape,
```
Comment (mattbahr) [inline tensorflow/python/ops/math_ops.py:4863]: Still not really too sure what to do about the `dense_shape` parameter here or if it's absolutely necessary

File: tensorflow/python/ops/math_ops.py (hunk 2 above)
Comment (mattbahr) [reply tensorflow/python/ops/math_ops.py]: Still need to do some more testing with multiple batch dimensions, but I removed this logic checking for 1 or no batch dims

File: tensorflow/python/ops/math_ops.py (hunk 8 above)
Comment (cantonios) [reply tensorflow/python/ops/math_ops.py:4863]: It's not necessary for the implementation, but I might suggest keeping it for consistency with other sparse ops.   You could use it for shape validation.

Or have your input be a `SparseTensor` and return a `SparseTensor`.

File: tensorflow/python/ops/math_ops.py [hunk 9]
```diff
@@ -4857,6 +4857,42 @@ def sparse_segment_sum(
...
+#  res = gen_math_ops.sampled_addmm(
+#      indices,
+#      values,
+#      dense_shape,
+#      mat1,
+#      mat2,
+#      beta=beta,
+#      alpha=alpha
+#  )
+  indices_dims = indices.get_shape().ndims
```
[the remark this answers is not in the document]
Comment (mattbahr) [reply tensorflow/python/ops/math_ops.py:4878]: Ok, thanks for the explanation! I think I'll close this PR and cut a clean branch with this implementation

**Pull request closed without merging**